contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    /**
+     * @dev Gets and verifies a warp out of band message, and adds the new protocol version
+     * addres to the registry.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Requirements:
+     *
+     * - a valid warp out of band message must be provided.
+     * - the version must be the increment of the latest version.
+     * - the protocol address must be a contract address.
+     */
+    function addProtocolVersion(uint32 messageIndex) external {
+        _addProtocolVersion(messageIndex);
+    }
+
+    /**
+     * @dev Gets the address of a protocol version.
+     * Requirements:
+     *
+     * - the version must be a valid version.
+     */
+    function getVersionToAddress(
+        uint256 version
+    ) external view returns (address) {
+        return _getVersionToAddress(version);
+    }
+
+    /**
+     * @dev Gets the latest protocol version.
+     */
+    function getLatestVersion() external view returns (uint256) {
+        return _latestVersion;
+    }
+
+    /**
+     * @dev Gets and verifies for a warp out of band message, and adds the new protocol version
+     * addres to the registry.
+     */
+    function _addProtocolVersion(uint32 messageIndex) internal virtual {
+        // Get and verify a valid warp out of band message.
+        (WarpMessage memory message, bool valid) = WARP_MESSENGER
+            .getVerifiedWarpMessage(messageIndex);
+        if (!valid) {
+            revert InvalidWarpMessage();
+        }
+        if (message.sourceChainID != _chainID) {
+            revert InvalidSourceChainID();
+        }
+        if (message.originSenderAddress != VALIDATORS_SOURCE_ADDRESS) {
+            revert InvalidOriginSenderAddress();
+        }
+        if (message.destinationChainID != _chainID) {
+            revert InvalidDestinationChainID();
+        }
+        if (message.destinationAddress != address(this)) {
+            revert InvalidDestinationAddress();
+        }
+
+        (uint256 version, address protocolAddress) = abi.decode(
+            message.payload,
+            (uint256, address)
+        );
+
+        _addToRegistry(version, protocolAddress);
+    }
+
+    /**
+     * @dev Adds the new protocol version address to the registry.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Requirements:
+     *
+     * - `version` must be the increment of the latest version.
+     * - `protocolAddress` must be a contract address.
+     */
+    function _addToRegistry(
+        uint256 version,
+        address protocolAddress
+    ) internal virtual {
+        // Check that the version is the increment of the latest version.
+        if (version != _latestVersion + 1) {
+            revert InvalidProtocolVersion();
+        }
+        // Check that the protocol address is a contract address.
+        if (!Address.isContract(protocolAddress)) {
+            revert InvalidProtocolAddress();
+        }
```

Comment:
I think it could potentially be useful to be able to add a protocol address prior to the contract being deployed. for instance, if a subnet has an important security patch to a protocol contract, they may want to register the new address prior to deploying the new contract code (which may make it clear what the vulnerability of the existing contract is).

Reply:
hmm yes that definitely sounds useful, will bring up to the team and see people's thoughts. I agree that for this case we'd probably want to remove the contract check, this was added to just help prevent a wrong protocol address from being specified, but if a significant threshold of the subnet agrees already shouldn't really be needed.

Reply:
removed the requirement for a contract address